cvelib/cve_api.py

```diff
@@ -87,29 +121,72 @@ def _post(self, path: str, **kwargs) -> requests.Response:
+            return cve_json["containers"]["cna"]
+        return cve_json
+
+    def _add_provider_metadata(self, cve_json: dict) -> dict:
+        """Add the providerMetadata objects to a CNA container if one is not present.
+
+        The orgId is the only required element of the providerMetadata object, and we can fetch
+        it from the API using the org short name provided by the user (when this class is
+        instantiated.)
+        """
+        if "providerMetadata" not in cve_json:
+            org_id = self.show_org()["UUID"]
+            cve_json["providerMetadata"] = {"orgId": org_id}
+        return cve_json
+
+    def publish(self, cve_id: str, cve_json: dict, validate: bool = True) -> dict:
         """Publish a CVE from a JSON object representing the CNA container data."""
+        cve_json = self._extract_cna_container(cve_json)
+        cve_json = self._add_provider_metadata(cve_json)
+        if validate:
+            CveRecord.validate(cve_json, CveRecord.Schemas.CNA_PUBLISHED)
+
         cve_json = {"cnaContainer": cve_json}
         response = self._post(f"cve/{cve_id}/cna", json=cve_json)
         response.raise_for_status()
         return response.json()
 
-    def update_published(self, cve_id: str, cve_json: dict) -> dict:
+    def update_published(self, cve_id: str, cve_json: dict, validate: bool = True) -> dict:
         """Update a published CVE record from a JSON object representing the CNA container data."""
+        cve_json = self._extract_cna_container(cve_json)
+        cve_json = self._add_provider_metadata(cve_json)
+        if validate:
+            CveRecord.validate(cve_json, CveRecord.Schemas.CNA_PUBLISHED)
+
         cve_json = {"cnaContainer": cve_json}
         response = self._put(f"cve/{cve_id}/cna", json=cve_json)
         response.raise_for_status()
         return response.json()
 
-    def reject(self, cve_id: str, cve_json: dict) -> dict:
+    def reject(self, cve_id: str, cve_json: dict, validate: bool = True) -> dict:
         """Reject a CVE from a JSON object representing the CNA container data."""
+        cve_json = self._extract_cna_container(cve_json)
+        cve_json = self._add_provider_metadata(cve_json)
+        if validate:
+            CveRecord.validate(cve_json, CveRecord.Schemas.CNA_REJECTED)
+
         cve_json = {"cnaContainer": cve_json}
         response = self._post(f"cve/{cve_id}/reject", json=cve_json)
         response.raise_for_status()
         return response.json()
 
-    def update_rejected(self, cve_id: str, cve_json: dict) -> dict:
+    def update_rejected(self, cve_id: str, cve_json: dict, validate: bool = True) -> dict:
         """Update a rejected CVE record from a JSON object representing the CNA container data."""
+        cve_json = self._extract_cna_container(cve_json)
+        cve_json = self._add_provider_metadata(cve_json)
+        if validate:
+            CveRecord.validate(cve_json, CveRecord.Schemas.CNA_REJECTED)
+
```

Comment:
Since this is repeated four times, can we make it into a clean_and_validate method?
```
    def clean_and_validate(cve_json: dict, validate: bool = True, schema_path: str = CveRecord.Schemas.CNA_PUBLISHED)
        cve_json = self._extract_cna_container(cve_json)
        cve_json = self._add_provider_metadata(cve_json)
        if validate:
            CveRecord.validate(cve_json, schema_path)
        return cve_json
```

Reply:
Eh, I'll plead https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction once again here :stuck_out_tongue: